### ISSUE TITLE
`vector_create_no_saves` doesn't return anything

### ISSUE THREAD

**Issue by nbansari** [nibblebits/DragonCompiler#78]
My C is a bit rusty, but in the vector_create_no_saves in [vector.c](https://github.com/nibblebits/DragonCompiler/blob/30f5f5898964ca91e6117c69d3293c8ea6f8a757/helpers/vector.c)
shouldn't the initialised vector be returned?

```c
struct vector *vector_create_no_saves(size_t esize)
{
    struct vector *vector = calloc(sizeof(struct vector), 1);
    vector->data = malloc(esize * VECTOR_ELEMENT_INCREMENT);
    vector->mindex = VECTOR_ELEMENT_INCREMENT;
    vector->rindex = 0;
    vector->pindex = 0;
    vector->esize = esize;
    vector->count = 0;
}
```

**Comment by nibblebits** [comment]
This particular function is not used within our compiler. These vector files I created many months ago to be used in several projects. Yes its a mistake in the Vector file but we don't use this function in the compiler at all. Solution is to delete the entire function which I will do

**Comment by nibblebits** [comment]
Okay it is actually used in vector_create but we dont use the save functionality. Yes this should be returned I will fix it.

**Comment by nibblebits** [comment]
This has been fixed thanks for your help: https://github.com/nibblebits/DragonCompiler/commit/082d6bed9bbb59af989d7a707060eceee1567809

**Comment by nbansari** [comment]
First off, thank you for the quick response.

I actually meant to post this as an issue for the PeachCompiler. 
**(You probably want to fix the issue there as well.)**

I came across your Udemy course which led me to this repo. 
I'm just curious (for my own understanding) to know why the compiler would function at all in spite of this bug, given vector_create is used throughout the code? Is it because of some undefined behaviour (like if the address of the created vector was in the return register)?

**Comment by nibblebits** [comment]
Hi Nban,
It has to be that some how the stack memory had an address we had access too, therefore allowing it to work. The actual structure is very small in size, therefore its understandable why it could work without a crash but as you say its undefined behavior. I also did make a new lecture fixing the problem which is in Module 3 of the course. However it will affect an incredibly small amount of people because as you see on the lecture videos we copy directly from this repository which has now been fixed.

The only time undefined behavior is possible is when we use vector_save which does not happen until the preprocessor stage of the course which is in Module 3. Since the bug has been fixed this issue will never arise.

Thanks,
Dan
